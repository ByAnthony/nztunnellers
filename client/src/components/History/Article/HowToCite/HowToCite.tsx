import STYLES from '../Notes/Notes.module.scss';

type Props = {
  title: string;
  today: Date;
};

export function HowToCite({ title, today }: Props) {
  return (
    <div className={STYLES.notes}>
      <h3>How to cite this page</h3>
      <p>
        Anthony Byledbal, &ldquo;
        {title.replace('\\', ' ')}
        &rdquo;,
        {' '}
        <em>New Zealand Tunnellers Website</em>
        ,
        {' '}
        {`${today.getFullYear()}`}
        {' '}
        (2009),
        Accessed:
        {' '}
        {`${today.toLocaleDateString('en-NZ', {
          year: 'numeric',
          month: 'long',
          day: 'numeric',
        })}. `}
      </p>
    </div>
  );
}
