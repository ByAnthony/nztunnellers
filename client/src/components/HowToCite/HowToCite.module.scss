@use 'src/index.scss';

.howtocite {
  margin-top: index.$base-margin;

  p {
    @include index.notes-paragraph
  }

  @include index.desktop {
    margin: 0 auto;
    padding: index.$xs-margin 0;
    max-width: 42rem;
  }
}
