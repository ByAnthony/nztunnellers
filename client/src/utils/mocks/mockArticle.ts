import { AboutUs } from '../../types/about-us';
import {
  Article, Image, Next, Section,
} from '../../types/article';

export const mockTitle: string = 'My Awesome\\Article Title';

export const mockChapter: number = 1;

export const mockTopImage: Image = {
  file: 'img-123.png',
  alt: 'Accessible alt text',
};

export const mockArticleImage: Image = {
  ...mockTopImage,
  title: 'An Amazing Photo',
  photographer: 'The French Photographer Doisneau',
  reference: 'Where the photo is preserved',
};

export const mockSectionOne: Section = {
  title: 'Section Title',
  text: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit,sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea *commodo consequat*. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.\\n\\nLorem ipsum dolor sit amet, consectetur adipiscing elit,sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat[1](#footnote_1). Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.\\n\\n(6--July 2023)',
};

export const mockSectionTwo: Section = {
  title: 'Section Title',
  text: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit,sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea *commodo consequat*. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.\\n\\nLorem ipsum dolor sit amet, consectetur adipiscing elit,sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat[1](#footnote_1). Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.\\n\\n(6--July 2023)',
};

export const mockNextButton: Next = {
  url: 'my-path-to-next-chapter/',
  title: 'Next\\Chapter',
  chapter: 3,
};

export const mockNotes: string = '[1.](#reference_1) Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.\\n\\n[2.](#reference_2) Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.';

export const mockArticle: Article = {
  id: 'my-awesome-article',
  chapter: mockChapter,
  title: mockTitle,
  section: [mockSectionOne, mockSectionTwo],
  image: [mockTopImage, mockArticleImage],
  next: mockNextButton,
  notes: mockNotes,
};

export const mockAboutUs: AboutUs = {
  id: 'my-awesome-article',
  title: mockTitle,
  section: [mockSectionOne, mockSectionTwo],
  image: [mockTopImage],
};
