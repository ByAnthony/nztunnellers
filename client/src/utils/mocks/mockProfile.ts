// import { Profile } from '../../types/tunneller';
// import { mockSummary } from './mockSummary';

export const mockId = 26;

// export const mockProfile = (overrides: Partial<Profile> | undefined = undefined): Profile => ({
//   id: mockId,
//   summary: mockSummary(),
//   origins: mockOrigins(),
//   preWarYears: mockPreWarYears(),
//   militaryYears: mockMilitaryYears(),
//   postServiceYears: mockPostServiceYears(),
//   sources: mockSources(),
//   image: mockImage(),
//   ...overrides,
// });
