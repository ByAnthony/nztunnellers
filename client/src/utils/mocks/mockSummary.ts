import { Summary } from '../../types/tunneller';

export const mockSummary: Summary = {
  serial: '1/2345',
  name: {
    forename: 'John',
    surname: 'Doe',
  },
  birth: '1888',
  death: '1975',
};
