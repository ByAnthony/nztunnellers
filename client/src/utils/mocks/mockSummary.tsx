import { Summary } from '../../types/tunneller';

export const testSummary: Summary = {
  serial: '12345',
  name: {
    forename: 'John',
    surname: 'Doe',
  },
  birth: '1888',
  death: '1975',
};
