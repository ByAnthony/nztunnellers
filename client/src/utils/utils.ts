export const displayBirthDeathDates = (
  birth: string | null,
  death: string | null,
) => {
  if (birth !== null && death !== null) {
    return `${birth} - ${death}`;
  }
  if (birth !== null && death === null) {
    return `${birth} - †?`;
  }
  if (birth === null && death !== null) {
    return `? - ${death}`;
  }
  return '? - †?';
};
