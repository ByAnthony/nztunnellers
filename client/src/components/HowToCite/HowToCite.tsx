import { displayBirthDeathDates } from '../../utils/utils';
import type { Summary } from '../../types/tunneller';
import STYLES from './HowToCite.module.scss';

type HowToCiteUrlProps = {
  id: number | undefined;
  title: string | undefined;
};

function HowToCiteUrl({ id, title }: HowToCiteUrlProps) {
  if (id) {
    return <span>{`URL: www.nztunnellers.com/tunnellers/${id}.`}</span>;
  }
  const articleTitle = title?.replace('\\', '-').replace(' ', '-').toLowerCase();
  return <span>{`URL: www.nztunnellers.com/history/${articleTitle}.`}</span>;
}

type HowToCiteTitleProps = {
  tunneller: Summary | undefined;
  title: string | undefined;
};

function HowToCiteTitle({ tunneller, title }: HowToCiteTitleProps) {
  if (tunneller) {
    return (
      <>
        {`${tunneller.name.forename} ${tunneller.name.surname} `}
        {`(${displayBirthDeathDates(tunneller.birth, tunneller.death)})`}
      </>
    );
  }
  const articleTitle = title?.replace('\\', ' ');
  return <span>{articleTitle}</span>;
}

type Props = {
  id?: number;
  summary?: Summary;
  title?: string;
  today: Date;
};

export function HowToCite({
  id, summary, title, today,
}: Props) {
  return (
    <div className={STYLES.howtocite}>
      <h3>How to cite this page</h3>
      <p>
        Anthony Byledbal, &ldquo;
        <HowToCiteTitle tunneller={summary} title={title} />
        &ldquo;,
        <em> New Zealand Tunnellers Website</em>
        {`, ${today.getFullYear()} (2009), Accessed: `}
        {`${today.toLocaleDateString('en-NZ', {
          year: 'numeric',
          month: 'long',
          day: 'numeric',
        })}. `}
        <HowToCiteUrl id={id} title={title} />
      </p>
    </div>
  );
}

HowToCite.defaultProps = {
  id: null,
  summary: null,
  title: null,
};
