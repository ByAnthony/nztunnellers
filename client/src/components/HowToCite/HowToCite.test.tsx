import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import { mockTitle } from '../../utils/mocks/mockArticle';
import { mockId } from '../../utils/mocks/mockProfile';
import { mockSummary } from '../../utils/mocks/mockSummary';
import { HowToCite } from './HowToCite';
import { findElementWithText } from '../../utils/test-utils';
import { mockToday } from '../../utils/mocks/mockToday';

describe('HowToCite for Profile', () => {
  const component = <HowToCite id={mockId} summary={mockSummary} today={mockToday} />;

  test('renders the component correctly', () => {
    const { asFragment } = render(component);

    expect(asFragment()).toMatchSnapshot();
  });

  describe('Name', () => {
    test('renders name correctly', () => {
      render(component);

      const element = findElementWithText('“John Doe (1888-1975)“');

      expect(element).toBeInTheDocument();
    });
  });

  describe('Dates', () => {
    test('renders birth and death dates when known', () => {
      render(component);

      const element = findElementWithText('“John Doe (1888-1975)“');

      expect(element).toBeInTheDocument();
    });

    test('renders only birth date when death date unknown', () => {
      const mockComponent = (
        <HowToCite
          id={mockId}
          summary={{ ...mockSummary, death: null }}
          today={mockToday}
        />
      );
      render(mockComponent);

      const element = findElementWithText('“John Doe (1888-†?)“');

      expect(element).toBeInTheDocument();
    });
  });

  describe('URL', () => {
    test('renders the url with correct id', () => {
      render(component);

      const element = findElementWithText('www.nztunnellers.com/tunnellers/26');

      expect(element).toBeInTheDocument();
    });
  });
});

describe('HowToCite for Article', () => {
  const component = (
    <HowToCite title={mockTitle} today={mockToday} />
  );

  test('renders the component correctly', () => {
    const { asFragment } = render(component);

    expect(asFragment()).toMatchSnapshot();
  });

  test('renders next chapter button correctly', () => {
    render(component);

    expect(screen.getByText(/How to cite this page/)).toBeInTheDocument();
    expect(screen.getByText(/My Awesome Article Title/)).toBeInTheDocument();
    expect(screen.getByText(/4 May 2023/)).toBeInTheDocument();
  });
});
