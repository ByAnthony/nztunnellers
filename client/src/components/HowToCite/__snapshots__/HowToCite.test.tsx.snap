// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HowToCite for Article renders the component correctly 1`] = `
<DocumentFragment>
  <div
    class="howtocite"
  >
    <h3>
      How to cite this page
    </h3>
    <p>
      Anthony Byledbal, “
      <span>
        My Awesome Article Title
      </span>
      “,
      <em>
         New Zealand Tunnellers Website
      </em>
      , 2023 (2009), Accessed: 4 May 2023. 
      <span>
        URL: www.nztunnellers.com/history/my-awesome-article-title.
      </span>
    </p>
  </div>
</DocumentFragment>
`;

exports[`HowToCite for Profile renders the component correctly 1`] = `
<DocumentFragment>
  <div
    class="howtocite"
  >
    <h3>
      How to cite this page
    </h3>
    <p>
      Anthony Byledbal, “John Doe (1888-1975)“,
      <em>
         New Zealand Tunnellers Website
      </em>
      , 2023 (2009), Accessed: 4 May 2023. 
      <span>
        URL: www.nztunnellers.com/tunnellers/26.
      </span>
    </p>
  </div>
</DocumentFragment>
`;
